### PR TITLE
Add variable sliding window CSV record reader

### PR DESCRIPTION
**Pull request by crockpotveggies**
## What changes were proposed in this pull request?

Adds support for sliding windows across CSV files. Windows are variable in length, according to index being read. User can also set "stride" of window.

## How was this patch tested?

See added tests. Currently passing.

### CODE REVIEW

File: datavec-api/src/main/java/org/datavec/api/records/reader/impl/csv/CSVVariableSlidingWindowRecordReader.java [hunk 1]
```diff
@@ -0,0 +1,235 @@
...
+import org.nd4j.linalg.primitives.Triple;
+
+import java.io.DataInputStream;
+import java.io.IOException;
+import java.net.URI;
+import java.util.*;
+
+/**
+ * A sliding window of variable size across an entire CSV.
+ *
```
Comment (AlexDBlack) [inline datavec-api/src/main/java/org/datavec/api/records/reader/impl/csv/CSVVariableSlidingWindowRecordReader.java]: Can we get a note on how the windowing works for the edges (start + end)?
From what I understand, it'll start with 1, then linearly increase to maxLinesPer sequence, then linearly decrease back to 1.

File: datavec-api/src/main/java/org/datavec/api/records/reader/impl/csv/CSVVariableSlidingWindowRecordReader.java [hunk 2]
```diff
@@ -0,0 +1,235 @@
...
+        this(maxLinesPerSequence, 0, stride, String.valueOf(CSVRecordReader.DEFAULT_DELIMITER));
+    }
+
+    /**
+     *
+     * @param maxLinesPerSequence    Number of lines in each sequences
+     * @param skipNumLines         Number of lines to skip at the start of the file (only skipped once, not per sequence)
+     * @param delimiter            Delimiter between entries in the same line, for example ","
+     */
+    public CSVVariableSlidingWindowRecordReader(int maxLinesPerSequence, int skipNumLines, int stride, String delimiter) {
```
Comment (AlexDBlack) [inline datavec-api/src/main/java/org/datavec/api/records/reader/impl/csv/CSVVariableSlidingWindowRecordReader.java]: Need Stride arg javadoc. Also probably want check on stride - make sure it's > 0

File: datavec-api/src/main/java/org/datavec/api/records/reader/impl/csv/CSVVariableSlidingWindowRecordReader.java [hunk 3]
```diff
@@ -0,0 +1,235 @@
...
+
+            if (queue.size() > maxLinesPerSequence || exhausted)
+                queue.poll();
+        }
+
+        for(List<Writable> line : queue) {
+            sequence.add(line);
+        }
+
+        Collections.reverse(sequence);
```
Comment (AlexDBlack) [inline datavec-api/src/main/java/org/datavec/api/records/reader/impl/csv/CSVVariableSlidingWindowRecordReader.java]: Why reverse the sequence? This doesn't seem correct to me...
(And if it is: I'd use queue.addFirst, then just do sequence = new ArrayList(queue)), insthead of the for loop + reverse)

File: datavec-api/src/test/java/org/datavec/api/records/reader/impl/CSVVariableSlidingWindowRecordReaderTest.java [hunk 4]
```diff
@@ -0,0 +1,80 @@
...
+import java.util.List;
+
+import static org.junit.Assert.assertEquals;
+
+/**
+ * Tests for variable sliding window csv reader.
+ *
+ * @author Justin Long (crockpotveggies)
+ */
+public class CSVVariableSlidingWindowRecordReaderTest {
```
Comment (AlexDBlack) [inline datavec-api/src/test/java/org/datavec/api/records/reader/impl/CSVVariableSlidingWindowRecordReaderTest.java]: Need better tests: notably,
(a) assert total number of sequences returned
(b) Test the metadata functionality

Comment (crockpotveggies) [inline datavec-api/src/test/java/org/datavec/api/records/reader/impl/CSVVariableSlidingWindowRecordReaderTest.java]: Fixed, and added test for strides.

File: datavec-api/src/main/java/org/datavec/api/records/reader/impl/csv/CSVVariableSlidingWindowRecordReader.java (hunk 1 above)
Comment (crockpotveggies) [inline datavec-api/src/main/java/org/datavec/api/records/reader/impl/csv/CSVVariableSlidingWindowRecordReader.java]: Fixed.

**Pull request merged**